strawberryfields/compilers/xunitary.py

```diff
@@ -103,22 +109,27 @@ def compile(self, seq, registers):
             # insert S2gates with 0 squeezing
             B.insert(0, Command(ops.S2gate(0, 0), [registers[i], registers[j]]))
 
-        sqs = [cmd.op.p[0] for cmd in B]
+        # get list of circuit registers as a tuple for each S2gate
+        regrefs = [(cmd.reg[0].ind, cmd.reg[1].ind) for cmd in B]
 
-        # ensure provided S2gates all have the allowed squeezing values
-        if not all(s in self.allowed_sq_ranges for s in sqs):
-            wrong_sq_values = [np.round(s, 4) for s in sqs if s not in self.allowed_sq_ranges]
-            raise CircuitError(
-                "Incorrect squeezing value(s) r={}. Allowed squeezing "
-                "value(s) are {}.".format(wrong_sq_values, self.allowed_sq_ranges)
-            )
-        # This could in principle be changed
-        phases = [cmd.op.p[1] for cmd in B]
-        if not np.allclose(phases, 0):
-            raise CircuitError(
-                "Incorrect phase value(s) phi={}. Allowed squeezing "
-                "value(s) are 0.0.".format(phases)
-            )
+        # merge S2gates
+        if len(regrefs) > half_n_modes:
+            for mode, indices in list_duplicates(regrefs):
+                r = 0
+                phi = 0
+
+                for k, i in enumerate(sorted(indices, reverse=True)):
+                    removed_cmd = B.pop(i)
+                    r += removed_cmd.op.p[0]
+                    phi_new = removed_cmd.op.p[1]
+
+                    if k > 0 and phi_new != phi:
+                        raise CircuitError("Cannot merge S2gates with different phase values.")
+
+                    phi = phi_new
+
+                i, j = mode
+                B.insert(indices[0], Command(ops.S2gate(r, phi), [registers[i], registers[j]]))
```

Comment:
@nquesada I noticed that your xunitary test file had a commented out test for merging S2 gates. I added the functionality here so that the test would pass

Reply:
I only consider the case where all S2gates to be merged have the same value of phi though